daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,272 @@
+        // normalization rule #2
+        makeRoll(m ++ List(t)) { norms =>
+          k(h :: norms)
+        }
+
+      case Case.Multi(h: Norm.Act, m, t: Norm.Roll) =>
+        // normalization rule #3
+        k(List(pushIntoRoll(h, m, t)))
+
+      case Case.Multi(h: Norm.Act, m, t: Norm.Act) =>
+        // no rule
+        k(List(Norm.Roll2(h, m, t)))
+    }
+  }
+
+  private def pushIntoRoll(a1: Norm.Act, xs2: List[Norm], t: Norm.Roll): Norm.Roll = {
+    t match {
+      case Norm.Roll1(a3) => Norm.Roll2(a1, xs2, a3)
+      case Norm.Roll2(a3, xs4, a5) => Norm.Roll2(a1, xs2 ++ List(a3) ++ xs4, a5)
+    }
+  }
+
+  // State manages a counter for node-id generation, and accumulates the nodes-map for
+  // the normalized transaction
+
+  // There is no connection between the ids in the original and normalized transaction.
+
+  private case class State(index: Int, nodeMap: Map[Nid, Node]) {
+
+    def next[R](k: (State, Nid) => Tramp[R]): Tramp[R] = {
+      k(State(index + 1, nodeMap), NodeId(index))
+    }
+
+    def push[R](nid: Nid, node: Node)(k: (State, Nid) => Tramp[R]): Tramp[R] = {
+      k(State(index, nodeMap = nodeMap + (nid -> node)), nid)
+    }
+
+  }
+
+  // The `push*` functions convert the Canonical types to the tx being collected in State.
+  // Ensuring:
+  // - The final tx has increasing node-ids when nodes are listed in pre-order.
+  // - The root node-id is 0 (we have tests that rely on this)
+
+  private val initialState = State(0, Map.empty)
+
+  private def pushAct[R](s: State, x: Norm.Act)(k: (State, Nid) => Tramp[R]): Tramp[R] = {
+    Bounce { () =>
+      s.next { (s, me) =>
+        x match {
+          case Norm.Leaf(node) =>
+            s.push(me, node)(k)
+          case Norm.Exe(exe, subs) =>
+            pushNorms(s, subs) { (s, children) =>
+              val node = exe.copy(children = ImmArray(children))
+              s.push(me, node)(k)
+            }
+        }
+      }
+    }
+  }
+
+  private def pushRoll[R](s: State, x: Norm.Roll)(k: (State, Nid) => Tramp[R]): Tramp[R] = {
+    s.next { (s, me) =>
+      x match {
+        case Norm.Roll1(act) =>
+          pushAct(s, act) { (s, child) =>
+            val node = NodeRollback(children = ImmArray(List(child)))
+            s.push(me, node)(k)
+          }
+        case Norm.Roll2(h, m, t) =>
+          pushAct(s, h) { (s, hh) =>
+            pushNorms(s, m) { (s, mm) =>
+              pushAct(s, t) { (s, tt) =>
+                val children = List(hh) ++ mm ++ List(tt)
+                val node = NodeRollback(children = ImmArray(children))
+                s.push(me, node)(k)
+              }
+            }
+          }
+      }
+    }
+  }
+
+  private def pushNorm[R](s: State, x: Norm)(k: (State, Nid) => Tramp[R]): Tramp[R] = {
+    x match {
+      case act: Norm.Act => pushAct(s, act)(k)
+      case roll: Norm.Roll => pushRoll(s, roll)(k)
+    }
+  }
+
+  private def pushNorms[R](s: State, xs: List[Norm])(
```

Comment:
I find the trampolined CPS style here pretty hard to follow. I hacked up a version of this that doesn’t use cps or trampolines and at least for me seems much easier to follow (ignore variable names and so on) https://github.com/digital-asset/daml/commit/2efd36a6f4efe69edbc2f48fd347cbd9b5ac916b. wdyt @remyhaemmerle-da @nickchapman-da? We can probably do something very similar for the first phase as well.

Reply:
I think the nice thing about this approach is that it's a systematic transformation of the original code.
See:
https://github.com/digital-asset/daml/pull/9712/commits/4511623e529c9c3d022538f3bf078ce6916d214b
https://github.com/digital-asset/daml/pull/9712/commits/e520a935e6943023d0f165b450f71c7f63077335

Reply:
We agreed to proceed with the trampoline approach, so long as Remy is happy. Which he is!

Reply:
I am fine with both approaches.